File: codes/python/ch07/cheb_matrix_structure.py
#!/usr/bin/env python3
"""
cheb_matrix_structure.py

Visualizes the structure of the Chebyshev differentiation matrix: heatmap
showing the matrix entries and row profile showing the decay pattern.

This script generates Figure 6.2 for Chapter 6.

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams
from matplotlib.colors import TwoSlopeNorm
from pathlib import Path

# Import Chebyshev matrix function
import sys
sys.path.insert(0, str(Path(__file__).parent))
from cheb_matrix import cheb_matrix

# -----------------------------------------------------------------------------
# Publication-quality matplotlib configuration
# -----------------------------------------------------------------------------
rcParams['font.family'] = 'serif'
rcParams['font.serif'] = ['Computer Modern Roman', 'CMU Serif', 'DejaVu Serif']
rcParams['mathtext.fontset'] = 'cm'
rcParams['axes.labelsize'] = 11
rcParams['axes.titlesize'] = 12
rcParams['xtick.labelsize'] = 10
rcParams['ytick.labelsize'] = 10
rcParams['legend.fontsize'] = 10
rcParams['figure.dpi'] = 150
rcParams['savefig.dpi'] = 300
rcParams['text.usetex'] = False
rcParams['axes.linewidth'] = 0.8
rcParams['xtick.major.width'] = 0.8
rcParams['ytick.major.width'] = 0.8

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
N = 16  # Number of intervals

# Book color scheme
NAVY = '#142D6E'
SKY = '#7896D2'
CORAL = '#E74C3C'
TEAL = '#16A085'
PURPLE = '#8E44AD'
ORANGE = '#E67E22'

# Output paths
SCRIPT_DIR = Path(__file__).parent
OUTPUT_DIR = SCRIPT_DIR / '..' / '..' / '..' / 'textbook' / 'figures' / 'ch07' / 'python'


def redblue_cmap():
    """Create a red-white-blue diverging colormap."""
    from matplotlib.colors import LinearSegmentedColormap
    colors = [(0.0, 0.0, 1.0),    # Blue
              (1.0, 1.0, 1.0),    # White
              (1.0, 0.0, 0.0)]    # Red
    return LinearSegmentedColormap.from_list('redblue', colors)


def main():
    """Create matrix structure visualization figure."""

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    # Construct differentiation matrix
    D, x = cheb_matrix(N)

    # Create figure with two panels
    fig, axes = plt.subplots(1, 2, figsize=(11, 5))

    # Panel 1: Heatmap of D matrix
    ax1 = axes[0]
    vmax = np.max(np.abs(D))

    # Use diverging colormap centered at zero
    im1 = ax1.imshow(D, cmap='RdBu_r', vmin=-vmax, vmax=vmax, aspect='equal')
    cbar = plt.colorbar(im1, ax=ax1, shrink=0.8)
    cbar.ax.set_ylabel('Matrix entry value', fontsize=10)

    ax1.set_title(f'Chebyshev Differentiation Matrix $D_N$ ($N = {N}$)', fontsize=11)
    ax1.set_xlabel('Column index $j$')
    ax1.set_ylabel('Row index $i$')

    # Set tick labels
    tick_positions = [0, N//4, N//2, 3*N//4, N]
    ax1.set_xticks(tick_positions)
    ax1.set_yticks(tick_positions)
    ax1.set_xticklabels([str(i) for i in tick_positions])
    ax1.set_yticklabels([str(i) for i in tick_positions])

    # Add annotation about large corner entries
    ax1.annotate(f'$D_{{00}} = {D[0,0]:.1f}$', xy=(0, 0), xytext=(4, 4),
                 fontsize=9, color='white',
                 bbox=dict(boxstyle='round', facecolor=NAVY, alpha=0.8))
    ax1.annotate(f'$D_{{N,N}} = {D[N,N]:.1f}$', xy=(N, N), xytext=(N-5, N-3),
                 fontsize=9, color='white',
                 bbox=dict(boxstyle='round', facecolor=NAVY, alpha=0.8))

    # Panel 2: Row profile (first row and middle row)
    ax2 = axes[1]

    j_indices = np.arange(N + 1)

    # First row (boundary)
    markerline, stemlines, baseline = ax2.stem(j_indices, D[0, :], basefmt=' ',
                                                label=f'Row $i=0$ (boundary)')
    markerline.set_color(CORAL)
    markerline.set_marker('o')
    stemlines.set_color(CORAL)

    # Middle row
    mid_row = N // 2
    markerline2, stemlines2, baseline2 = ax2.stem(j_indices + 0.15, D[mid_row, :],
                                                   basefmt=' ', label=f'Row $i={mid_row}$ (interior)')
    markerline2.set_color(TEAL)
    markerline2.set_marker('s')
    stemlines2.set_color(TEAL)

    ax2.axhline(0, color='#888888', linewidth=0.5)
    ax2.set_xlabel('Column index $j$')
    ax2.set_ylabel('Matrix entry $D_{ij}$')
    ax2.set_title('Row Profiles of $D_N$', fontsize=11)
    ax2.legend(loc='upper right', fontsize=9)
    ax2.spines['top'].set_visible(False)
    ax2.spines['right'].set_visible(False)

    # Add grid
    ax2.yaxis.grid(True, linestyle='-', alpha=0.2, color='#888888')
    ax2.set_xlim(-0.5, N + 0.5)

    plt.tight_layout()

    # Save figure
    output_file = OUTPUT_DIR / 'cheb_matrix_structure.pdf'
    fig.savefig(output_file, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(output_file.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)

    print(f'Figure saved to: {output_file.resolve()}')
    plt.close(fig)

    # Print matrix properties
    print(f"\nChebyshev Matrix Properties (N = {N}):")
    print("-" * 50)
    print(f"  Matrix size: ({N+1}, {N+1})")
    print(f"  Corner entry D[0,0]: {D[0,0]:.6f}")
    print(f"  Corner entry D[N,N]: {D[N,N]:.6f}")
    print(f"  Theoretical D[0,0] = (2N^2+1)/6: {(2*N**2+1)/6:.6f}")
    print(f"  Max entry: {np.max(D):.6f}")
    print(f"  Min entry: {np.min(D):.6f}")
    print(f"  Negative sum trick error: {np.max(np.abs(D @ np.ones(N+1))):.2e}")
    print("-" * 50)


if __name__ == '__main__':
    main()

File: codes/julia/ch07/cheb_matrix_structure.jl
# cheb_matrix_structure.jl
#
# Visualizes the structure of the Chebyshev differentiation matrix: heatmap
# showing the matrix entries and row profile showing the decay pattern.
#
# This script generates Figure 6.2 for Chapter 6.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra

# Import Chebyshev matrix function
include(joinpath(@__DIR__, "cheb_matrix.jl"))

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const N = 16  # Number of intervals

# Book color scheme
const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#9B59B6"
const ORANGE = colorant"#E67E22"

# Output paths
SCRIPT_DIR = @__DIR__
OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch07", "julia")
mkpath(OUTPUT_DIR)

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Construct differentiation matrix
    D, x = cheb_matrix(N)

    # --- Figure 1: Heatmap + Row profiles ---
    fig = Figure(size = (800, 400))

    # Panel 1: Heatmap of D matrix
    ax1 = Axis(fig[1, 1],
               xlabel = L"Column index $j$",
               ylabel = L"Row index $i$",
               title = "Chebyshev Differentiation Matrix \$D_N\$ (\$N = $N\$)",
               yreversed = true,
               aspect = DataAspect())

    vmax = maximum(abs.(D))

    # Heatmap with diverging colormap centered at zero
    hm = heatmap!(ax1, 0:N, 0:N, D',
                  colormap = :RdBu, colorrange = (-vmax, vmax))
    Colorbar(fig[1, 1, Right()], hm, label = "Matrix entry value",
             labelsize = 10, width = 12)

    # Set ticks
    tick_positions = [0, N ÷ 4, N ÷ 2, 3 * N ÷ 4, N]
    ax1.xticks = (tick_positions, string.(tick_positions))
    ax1.yticks = (tick_positions, string.(tick_positions))

    # Annotations for corner entries
    text!(ax1, 4.0, 4.0,
          text = "\$D_{00} = $(round(D[1,1], digits=1))\$",
          fontsize = 9, color = :white, align = (:left, :center),
          font = "CMU Serif")
    text!(ax1, N - 5.0, N - 3.0,
          text = "\$D_{N,N} = $(round(D[N+1,N+1], digits=1))\$",
          fontsize = 9, color = :white, align = (:left, :center),
          font = "CMU Serif")

    # Panel 2: Row profiles (first row and middle row)
    ax2 = Axis(fig[1, 2],
               xlabel = L"Column index $j$",
               ylabel = L"Matrix entry $D_{ij}$",
               title = L"Row Profiles of $D_N$")
    hidespines!(ax2, :t, :r)

    j_indices = 0:N

    # First row (boundary)
    stemlines_x1 = Float64[]
    stemlines_y1 = Float64[]
    for j in j_indices
        push!(stemlines_x1, j); push!(stemlines_y1, 0.0)
        push!(stemlines_x1, j); push!(stemlines_y1, D[1, j + 1])
        push!(stemlines_x1, NaN); push!(stemlines_y1, NaN)
    end
    lines!(ax2, stemlines_x1, stemlines_y1, color = CORAL, linewidth = 1.0)
    scatter!(ax2, collect(Float64, j_indices), D[1, :],
             color = CORAL, markersize = 6, marker = :circle,
             label = L"Row $i=0$ (boundary)")

    # Middle row
    mid_row = N ÷ 2
    stemlines_x2 = Float64[]
    stemlines_y2 = Float64[]
    for j in j_indices
        push!(stemlines_x2, j + 0.15); push!(stemlines_y2, 0.0)
        push!(stemlines_x2, j + 0.15); push!(stemlines_y2, D[mid_row + 1, j + 1])
        push!(stemlines_x2, NaN); push!(stemlines_y2, NaN)
    end
    lines!(ax2, stemlines_x2, stemlines_y2, color = TEAL, linewidth = 1.0)
    scatter!(ax2, collect(Float64, j_indices) .+ 0.15, D[mid_row + 1, :],
             color = TEAL, markersize = 6, marker = :rect,
             label = "Row \$i=$mid_row\$ (interior)")

    # Reference line at zero
    hlines!(ax2, 0, color = RGBf(0.533, 0.533, 0.533), linewidth = 0.5)

    # Horizontal grid
    ax2.ygridvisible = true
    ax2.ygridcolor = RGBAf(0.533, 0.533, 0.533, 0.2)
    ax2.ygridstyle = :solid

    xlims!(ax2, -0.5, N + 0.5)
    axislegend(ax2, position = :rt, labelsize = 9, framevisible = false)

    # Save first figure
    save(joinpath(OUTPUT_DIR, "cheb_matrix_structure.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "cheb_matrix_structure.png"), fig, px_per_unit = 4)
    println("Figure saved to: $(joinpath(OUTPUT_DIR, "cheb_matrix_structure.pdf"))")

    # --- Figure 2: Interior row profile only ---
    fig2 = Figure(size = (500, 330))

    interior_row = 9  # 0-based index
    ax3 = Axis(fig2[1, 1],
               xlabel = L"Column index $j$",
               ylabel = L"Matrix entry $D_{9,j}$",
               title = L"Interior Row Profile ($i = 9$)")
    hidespines!(ax3, :t, :r)

    # Stem plot for row i=9
    stemlines_x3 = Float64[]
    stemlines_y3 = Float64[]
    for j in j_indices
        push!(stemlines_x3, j); push!(stemlines_y3, 0.0)
        push!(stemlines_x3, j); push!(stemlines_y3, D[interior_row + 1, j + 1])
        push!(stemlines_x3, NaN); push!(stemlines_y3, NaN)
    end
    lines!(ax3, stemlines_x3, stemlines_y3, color = TEAL, linewidth = 1.0)
    scatter!(ax3, collect(Float64, j_indices), D[interior_row + 1, :],
             color = TEAL, markersize = 8, marker = :rect)

    # Reference line at zero
    hlines!(ax3, 0, color = RGBf(0.533, 0.533, 0.533), linewidth = 0.5)

    # Horizontal grid
    ax3.ygridvisible = true
    ax3.ygridcolor = RGBAf(0.533, 0.533, 0.533, 0.2)
    ax3.ygridstyle = :solid

    xlims!(ax3, -0.5, N + 0.5)

    # Save second figure
    save(joinpath(OUTPUT_DIR, "cheb_matrix_interior_row.pdf"), fig2)
    save(joinpath(OUTPUT_DIR, "cheb_matrix_interior_row.png"), fig2, px_per_unit = 4)
    println("Interior row figure saved to: $(joinpath(OUTPUT_DIR, "cheb_matrix_interior_row.pdf"))")

    # Print matrix properties
    println("\nChebyshev Matrix Properties (N = $N):")
    println("-" ^ 50)
    println("  Matrix size: ($(N+1), $(N+1))")
    println("  Corner entry D[0,0]: $(round(D[1,1], digits=6))")
    println("  Corner entry D[N,N]: $(round(D[N+1,N+1], digits=6))")
    println("  Theoretical D[0,0] = (2N^2+1)/6: $(round((2*N^2+1)/6, digits=6))")
    println("  Max entry: $(round(maximum(D), digits=6))")
    println("  Min entry: $(round(minimum(D), digits=6))")
    println("  Negative sum trick error: $(round(maximum(abs.(D * ones(N+1))), sigdigits=2))")
    println("-" ^ 50)
end

main()
